File: Ahorn/entities/templeGateAllSwitches.jl
module FlushelineTempleGateAllSwitches

using ..Ahorn, Maple

@mapdef Entity "vitellary/templegateall" Gate(x::Integer, y::Integer, sprite::String="default")

const textures = String["default", "mirror", "theo"]

const placements = Ahorn.PlacementDict(
    "Temple Gate (Default, All Switches) (Crystalline)" => Ahorn.EntityPlacement(
        Gate,
		"point",
		Dict{String, Any}(
            "sprite" => "default"
        )
    )
)

Ahorn.editingOptions(entity::Gate) = Dict{String, Any}(
    "sprite" => textures
)

function Ahorn.selection(entity::Gate)
    x, y = Ahorn.position(entity)

    return Ahorn.Rectangle(x - 4, y, 15, 48)
end

const sprites = Dict{String, String}(
    "default" => "objects/door/TempleDoor00",
    "mirror" => "objects/door/TempleDoorB00",
    "theo" => "objects/door/TempleDoorC00"
)

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::Gate, room::Maple.Room)
    sprite = get(entity.data, "sprite", "default")

    if haskey(sprites, sprite)
        Ahorn.drawImage(ctx, sprites[sprite], -4, 0)
    else
		Ahorn.drawImage(ctx, "objects/door/TempleDoor00", -4, 0)
	end
end

end

File: Loenn/entities/temple_gate_all_switches.lua
local drawableSprite = require "structs.drawable_sprite"
local celesteEnums = require "consts.celeste_enums"
local utils = require "utils"

local templeGate = {}

local textures = {
    default = "objects/door/TempleDoor00",
    mirror = "objects/door/TempleDoorB00",
    theo = "objects/door/TempleDoorC00"
}

local textureOptions = {}
for texture, _ in pairs(textures) do
    textureOptions[utils.titleCase(texture)] = texture
end

templeGate.name = "vitellary/templegateall"
templeGate.depth = -9000
templeGate.fieldInformation = {
    sprite = {
        options = textureOptions,
        editable = false
    },
}
templeGate.placements = {
    {
        name = "default",
        data = {
            sprite = "default",
        },
    },
}

function templeGate.sprite(room, entity)
    local variant = entity.sprite or "default"
    local texture = textures[variant] or textures["default"]
    local sprite = drawableSprite.fromTexture(texture, entity)

    -- Weird offset from the code, justifications are from sprites.xml
    sprite:setJustification(0.5, 0.0)
    sprite:addPosition(4, 0)

    return sprite
end

return templeGate